src/omv/ports/mimxrt/omv_spi.c

```diff
@@ -78,65 +89,118 @@ static void spi_master_callback(LPSPI_Type *base, void *handle, status_t status,
+            spi->xfer_descr.txData += offset;
+        }
+        if (spi->xfer_descr.rxData) {
+            spi->xfer_descr.rxData += offset;
         }
+        spi->xfer_flags ^= OMV_SPI_XFER_HALF;
+    }
+
+    // Start the next DMA transfer before calling the callback. This minimizes the time
+    // when DMA is not running. Also, if the callback aborts this will stop the DMA transfer.
+    if ((spi->dma_flags & OMV_SPI_DMA_CIRCULAR) && ((spi->xfer_flags & flags) == flags)) {
+        // Restart transfer for circular transfers. Note that we can't be interrupted again
+        // by this until this callback finishes so it is okay to clear xfer complete later.
+        LPSPI_MasterTransferEDMA(spi->inst, &spi->descr_master_edma, &spi->xfer_descr);
     }
 
     if (spi->callback) {
-        spi->callback(spi, spi->userdata);
+        spi->callback(spi, spi->userdata, buf);
     }
 
-    uint32_t flags = (OMV_SPI_XFER_DMA | OMV_SPI_XFER_CIRCULAR | OMV_SPI_XFER_COMPLETE);
-    if ((spi->xfer_flags & flags) == flags) {
-        // Restart transfer for circular transfers.
-        spi->xfer_flags &= ~(OMV_SPI_XFER_COMPLETE);
-        LPSPI_MasterTransferEDMA(spi->inst, &spi->descr_master_edma, &spi->xfer_descr);
+    // Clear after the callback so the callback gets the xfer complete flag set.
+    // This needs to be cleared to prevent circular DMA from re-triggering on a failure.
+    spi->xfer_flags &= ~(OMV_SPI_XFER_COMPLETE);
+}
+
+// The LPSPI_MasterTransferNonBlocking function call does not clear the LPSPI_TCR_FRAMESZ_MASK bit
+// even if kLPSPI_MasterByteSwap is not set in the xfer. So, we have to clear this bit at the same
+// time as changing the framesize. Splitting this will result in inconsitent settings. Finally,
+// the TCR register doesn't always update itself correctly so you have to try in a loop...
+void omv_spi_set_frame_size(omv_spi_t *spi, uint16_t size) {
+    for (;;) {
+        uint32_t tcr = spi->inst->TCR;
+        uint16_t s = (uint16_t) ((tcr & LPSPI_TCR_FRAMESZ_MASK) >> LPSPI_TCR_FRAMESZ_SHIFT) + 1U;
+        if (s != size) {
+            spi->inst->TCR = (tcr & ~(LPSPI_TCR_BYSW_MASK | LPSPI_TCR_FRAMESZ_MASK)) | LPSPI_TCR_FRAMESZ(size - 1U);
+        } else {
+            break;
+        }
     }
 }
 
 int omv_spi_transfer_start(omv_spi_t *spi, omv_spi_transfer_t *xfer) {
+    // No TX trasnfers in circular or double buffer mode.
+    if ((spi->dma_flags & (OMV_SPI_DMA_CIRCULAR | OMV_SPI_DMA_DOUBLE)) && xfer->txbuf) {
+        return -1;
+    }
+
     spi->callback = xfer->callback;
     spi->userdata = xfer->userdata;
+    spi->xfer_error = 0;
     spi->xfer_flags = xfer->flags;
-    // Duplicate & stash transfer in spi struct, to avoid recreating
-    // it for circular transfers. NOTE: If circular transfers ever
-    // works this can be removed.
+
+    uint32_t datasize = 8;
+    if (xfer->flags & OMV_SPI_XFER_16_BIT) {
+        datasize = 16;
+    }
+
+    omv_spi_set_frame_size(spi, datasize);
```

Comment:
Sorry this one is on me, but I thought about this some more and I prefer if we set the transfer/word width at config time, for the following reasons:

1- Writes to `TCR` take a number of SPI clock cycles to update, this will be done on * every * single transfer vs just once at config time.
2- This "feature" might be hard to support for some future port, if it is an issue for some port we'll end adding ugly code here that will deinit/init the SPI bus. I prefer if we keep this common API simple so it makes it easier to port it.
3- Most of the time you don't change the word width between transfers. The only code that does that is the LCD driver, it's a very special case forced on every use case (TV, WINC, PIXART, FLIR, etc..)
4- None of the SDKs/HALs etc support changing the word width in transmit/receive functions, the old code wrote to registers directly, so shouldn't try to support this either.

For the LCD driver, can we just reinit the bus when needed ?

Reply:
1. The driver code itself touches the register and changes it per API call. The only reason I have to touch it ourselves is that their code never updates the frame size. It just starts/stops the transaction. Note that I only write the register and delay in the loop if it's not what it should be.

2. Maybe... I don't see how you avoid the spi driver being super painful to make work. But, yes, if I had choosen to init/deinit per call I wouldn't have had to spend so much time learning why the SPI bus didn't work for random. It would have been way less painful. I can agree that getting down to the low-level details of the SPI device was challenging.

3. Sure, but, a new device may require this. Since we have the code working right now why not keep it? For future ports we can do the less effort init/deinit in the driver itself. I agree it would have saved time.

4. Yeah, but, you'd agree the HAL drivers aren't complete for every situation. We pretty much always have to bypass them for the best performance. E.g. the camera driver.

> For the LCD driver, can we just reinit the bus when needed ?

If we do that it should still be inside of the SPI driver right? Note, the LCD module will have to reinit and deinit the bus before and after every display update. I have to do a single byte write before updating the screen and then a single byte write afterwards.

Reply:
I didn't realize it required switching word width on every frame, but why can't you write the frame as bytes ? Which LCD is this for ?

Reply:
You can, but, but, then the processor has to byte reverse every 16-bits. So, it's a ton of CPU overhead. By just writing the data out at 16-bits at a time the processor can just fire off a DMA transaction from the RGB565 line buffer and go on it's way.

This for our normal LCD driver. I'm just leveraging how little-endian reads work to dramatically save CPU time. E.g. if the DMA reads data at 8-bits at a time it byte reverses the data versus if it reads it at 16-bits at a time.

Reply:
Updated the PR per feedback. However, I have not retested everything yet. Do not merge.

Reply:
I think the LCD driver had a bit to reverse the order.

Reply:
Yes, I removed that years ago when I optimized it.

Reply:
>  Yes, I removed that years ago when I optimized it.

Please un-optimize it, so we can fix this issue.